core/src/test/java/nu/yona/server/util/LockPoolTest.java

```diff
@@ -0,0 +1,165 @@
+	{
+		LockPool<Integer> lockPool = new LockPool<>();
+		ExecutorService threadPool = Executors.newFixedThreadPool(numThreads);
+		CountDownLatch startSignal = new CountDownLatch(1);
+		CountDownLatch doneSignal = new CountDownLatch(numThreads);
+		AtomicBoolean concurrencyIndicator = new AtomicBoolean(false);
+		AtomicInteger failureCounter = new AtomicInteger(0);
+		for (int i = 0; (i < numThreads); i++)
+		{
+			threadPool.execute(() -> tryAccess(iterations, startSignal, lockPool, concurrencyIndicator, failureCounter,
+					doneSignal, accessAttempt));
+		}
+		startSignal.countDown();
+		awaitWithoutInterrupt(doneSignal);
+		threadPool.shutdown();
+		return failureCounter.get();
+	}
+
+	private void tryAccess(int iterations, CountDownLatch startSignal, LockPool<Integer> lockPool,
+			AtomicBoolean concurrencyIndicator, AtomicInteger failureCounter, CountDownLatch doneSignal,
+			AccessAttempt accessAtempt)
+	{
+		awaitWithoutInterrupt(startSignal);
+		for (int i = 0; (i < iterations); i++)
+		{
+			accessAtempt.attempt(lockPool, concurrencyIndicator, failureCounter);
+		}
+		doneSignal.countDown();
+	}
+
+	private void attemptNonReentrantAccess(LockPool<Integer> lockPool, AtomicBoolean concurrencyIndicator,
+			AtomicInteger failureCounter)
+	{
+		int lockID = 0;
+		try (LockPool<Integer>.Lock lock = lockPool.lock(lockID))
+		{
+			verifyNonConcurrentExecution(concurrencyIndicator, failureCounter);
+		}
+	}
+
+	private void attemptReentrantAccess(LockPool<Integer> lockPool, AtomicBoolean concurrencyIndicator,
+			AtomicInteger failureCounter)
+	{
+		int lockID = 0;
+		try (LockPool<Integer>.Lock lock1 = lockPool.lock(0))
+		{
+			verifyNonConcurrentExecution(concurrencyIndicator, failureCounter);
+			try (LockPool<Integer>.Lock lock2 = lockPool.lock(lockID))
+			{
+				verifyNonConcurrentExecution(concurrencyIndicator, failureCounter);
+			}
+			verifyNonConcurrentExecution(concurrencyIndicator, failureCounter);
+		}
+	}
+
+	private void attempImmediateAccessRandomID(LockPool<Integer> lockPool, AtomicBoolean concurrencyIndicator,
+			AtomicInteger failureCounter)
+	{
+		int lockID = freeLockID.incrementAndGet();
+		int sleepDuration = 500;
+		long startTime = System.currentTimeMillis();
+		long accessTime;
+		try (LockPool<Integer>.Lock lock1 = lockPool.lock(lockID))
+		{
+			try (LockPool<Integer>.Lock lock2 = lockPool.lock(lockID))
+			{
+				accessTime = System.currentTimeMillis();
+				sleepWithoutInterrupt(sleepDuration);
+			}
+		}
+		if (accessTime - startTime > sleepDuration / 2)
+		{
+			// Delay was more than half the sleep duration, so we apparently were held up
+			failureCounter.incrementAndGet();
+		}
+	}
+
+	private void verifyNonConcurrentExecution(AtomicBoolean concurrencyIndicator, AtomicInteger failureCounter)
+	{
+		if (concurrencyIndicator.compareAndSet(false, true))
+		{
+			// No one else is here
+			concurrencyIndicator.set(false);
```

Comment:
Yes, I've thought of that and didn't like it because it slows down the test. When it's not necessary, I don't like to add sleeps. At the other hand, 1ms is very little. I'll give it a try and see what happens